pkg/pgmodel/sql_ingest.go

```diff
@@ -522,34 +573,47 @@ func tryRecovery(conn pgxConn, req copyRequest, err error) error {
 	return err
 }
 
-func doInsert(conn pgxConn, req copyRequest) (err error) {
-	numRows := 0
-	for i := range req.data.batch.sampleInfos {
-		numRows += len(req.data.batch.sampleInfos[i].samples)
-	}
-	times := make([]time.Time, 0, numRows)
-	vals := make([]float64, 0, numRows)
-	series := make([]int64, 0, numRows)
-	for req.data.batch.Next() {
-		time, val, serie := req.data.batch.Values()
-		times = append(times, time)
-		vals = append(vals, val)
-		series = append(series, int64(serie))
-	}
-	if len(times) != numRows {
-		panic("invalid insert request")
+func doInsert(conn pgxConn, reqs ...copyRequest) (err error) {
+	batch := conn.NewBatch()
+	numRowsPerInsert := make([]int, 0, len(reqs))
+	for _, req := range reqs {
+		numRows := 0
+		for i := range req.data.batch.sampleInfos {
+			numRows += len(req.data.batch.sampleInfos[i].samples)
+		}
+		times := make([]time.Time, 0, numRows)
+		vals := make([]float64, 0, numRows)
+		series := make([]int64, 0, numRows)
+		for req.data.batch.Next() {
+			time, val, serie := req.data.batch.Values()
+			times = append(times, time)
+			vals = append(vals, val)
+			series = append(series, int64(serie))
+		}
+		if len(times) != numRows {
+			panic("invalid insert request")
```

Comment:
I hope, this isn't too much of a blocking code, what it might lead to sample gaps when prometheus queries most recently written samples.